core/services/ocr2/plugins/ocr2keeper/evm/registry.go

```diff
@@ -0,0 +1,753 @@
+	keeper_registry_wrapper2_0.KeeperRegistryUpkeepCanceled{}.Topic(),
+	keeper_registry_wrapper2_0.KeeperRegistryUpkeepPaused{}.Topic(),
+	keeper_registry_wrapper2_0.KeeperRegistryUpkeepUnpaused{}.Topic(),
+	keeper_registry_wrapper2_0.KeeperRegistryFundsAdded{}.Topic(),
+}
+
+var upkeepActiveEvents = []common.Hash{
+	keeper_registry_wrapper2_0.KeeperRegistryUpkeepPerformed{}.Topic(),
+}
+
+type checkResult struct {
+	ur  []types.UpkeepResult
+	err error
+}
+
+type activeUpkeep struct {
+	ID              *big.Int
+	PerformGasLimit uint32
+	CheckData       []byte
+}
+
+type EvmRegistry struct {
+	HeadWatcher
+	sync          utils.StartStopOnce
+	poller        logpoller.LogPoller
+	addr          common.Address
+	client        client.Client
+	registry      *keeper_registry_wrapper2_0.KeeperRegistry
+	abi           abi.ABI
+	packer        *evmRegistryPackerV2_0
+	chLog         chan logpoller.Log
+	reInit        *time.Timer
+	mu            sync.RWMutex
+	filterID      int
+	lastPollBlock int64
+	ctx           context.Context
+	cancel        context.CancelFunc
+	active        map[int64]activeUpkeep
+	headFunc      func(types.BlockKey)
+	runState      int
+	runError      error
+}
+
+// GetActiveUpkeepKeys uses the latest head and map of all active upkeeps to build a
+// slice of upkeep keys.
+func (r *EvmRegistry) GetActiveUpkeepKeys(context.Context, types.BlockKey) ([]types.UpkeepKey, error) {
+	if r.LatestBlock() == 0 {
+		return nil, fmt.Errorf("%w: service probably not yet started", ErrHeadNotAvailable)
+	}
+
+	r.mu.RLock()
+	keys := make([]types.UpkeepKey, len(r.active))
+	var i int
+	for _, value := range r.active {
+		keys[i] = blockAndIdToKey(big.NewInt(r.LatestBlock()), value.ID)
+		i++
+	}
+	r.mu.RUnlock()
+
+	return keys, nil
+}
+
+func (r *EvmRegistry) CheckUpkeep(ctx context.Context, keys ...types.UpkeepKey) (types.UpkeepResults, error) {
+	chResult := make(chan checkResult, 1)
+	go r.doCheck(ctx, keys, chResult)
+
+	select {
+	case rs := <-chResult:
+		return rs.ur, rs.err
+	case <-ctx.Done():
+		// safety on context done to provide an error on context cancellation
+		// contract calls through the geth wrappers are a bit of a black box
+		// so this safety net ensures contexts are fully respected and contract
+		// call functions have a more graceful closure outside the scope of
+		// CheckUpkeep needing to return immediately.
+		return nil, fmt.Errorf("%w: failed to check upkeep on registry", ErrContextCancelled)
+	}
+}
+
+func (r *EvmRegistry) IdentifierFromKey(key types.UpkeepKey) (types.UpkeepIdentifier, error) {
+	_, id, err := blockAndIdFromKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return id.Bytes(), nil
+}
+
+func (r *EvmRegistry) Start(ctx context.Context) error {
+	return r.sync.StartOnce("AutomationRegistry", func() error {
+		r.mu.Lock()
+		defer r.mu.Unlock()
+		r.ctx, r.cancel = context.WithCancel(context.Background())
+		r.reInit = time.NewTimer(reInitializationDelay)
```

Comment:
Let's make the re-initialiation independent of logs. we can use 5m or something as time interval